File: data_structure/heap/src/HeapDemo.swift
// Swift
// データ構造: ヒープ (Heap)

import Foundation

class HeapData {
    private var data: [Int]
    private let isMinHeap: Bool
    
    init(isMinHeap: Bool = true) {
        self.data = []
        self.isMinHeap = isMinHeap
    }
    
    private func getParentIndex(_ index: Int) -> Int {
        return (index - 1) / 2
    }
    
    private func getLeftChildIndex(_ index: Int) -> Int {
        return 2 * index + 1
    }
    
    private func getRightChildIndex(_ index: Int) -> Int {
        return 2 * index + 2
    }
    
    private func hasParent(_ index: Int) -> Bool {
        return getParentIndex(index) >= 0
    }
    
    private func hasLeftChild(_ index: Int) -> Bool {
        return getLeftChildIndex(index) < data.count
    }
    
    private func hasRightChild(_ index: Int) -> Bool {
        return getRightChildIndex(index) < data.count
    }
    
    private func shouldSwap(_ index1: Int, _ index2: Int) -> Bool {
        if isMinHeap {
            return data[index1] > data[index2]
        } else {
            return data[index1] < data[index2]
        }
    }
    
    private func swap(_ index1: Int, _ index2: Int) {
        data.swapAt(index1, index2)
    }
    
    private func heapifyDown(_ index: Int) {
        var currentIndex = index
        var smallestOrLargest = currentIndex
        
        if hasLeftChild(currentIndex) && shouldSwap(smallestOrLargest, getLeftChildIndex(currentIndex)) {
            smallestOrLargest = getLeftChildIndex(currentIndex)
        }
        
        if hasRightChild(currentIndex) && shouldSwap(smallestOrLargest, getRightChildIndex(currentIndex)) {
            smallestOrLargest = getRightChildIndex(currentIndex)
        }
        
        if smallestOrLargest != currentIndex {
            swap(currentIndex, smallestOrLargest)
            heapifyDown(smallestOrLargest)
        }
    }
    
    private func heapifyUp(_ index: Int) {
        var currentIndex = index
        while hasParent(currentIndex) && shouldSwap(getParentIndex(currentIndex), currentIndex) {
            let parentIndex = getParentIndex(currentIndex)
            swap(parentIndex, currentIndex)
            currentIndex = parentIndex
        }
    }
    
    func get() -> [Int] {
        return data
    }
    
    func getIndex(of item: Int) -> Int {
        return data.firstIndex(of: item) ?? -1
    }
    
    func getValue(at index: Int) -> Int? {
        guard 0 <= index, index < data.count else {
            return nil
        }
        return data[index]
    }
    
    @discardableResult
    func heapify(_ array: [Int]) -> Bool {
        data = array
        for i in stride(from: data.count / 2 - 1, through: 0, by: -1) {
            heapifyDown(i)
        }
        return true
    }
    
    @discardableResult
    func push(_ value: Int) -> Bool {
        data.append(value)
        heapifyUp(data.count - 1)
        return true
    }
    
    @discardableResult
    func pop() -> Bool {
        guard !data.isEmpty else { return false }
        
        let lastElement = data.removeLast()
        
        if !data.isEmpty {
            data[0] = lastElement
            heapifyDown(0)
        }
        
        return true
    }
    
    func peek() -> Int? {
        return data.first
    }
    
    func isEmpty() -> Bool {
        return data.isEmpty
    }
    
    func size() -> Int {
        return data.count
    }
    
    @discardableResult
    func clear() -> Bool {
        data.removeAll()
        return true
    }
}

func main() {
    print("Heap TEST -----> start")
    
    print("\nmin heap: new")
    let minHeap = HeapData(isMinHeap: true)
    print("  現在のデータ: \(minHeap.get())")
    
    print("\nmin heap: heapify")
    let input = [4, 10, 3, 5, 1]
    print("  入力値: \(input)")
    let output = minHeap.heapify(input)
    print("  出力値: \(output)")
    print("  現在のデータ: \(minHeap.get())")
    
    print("\nmin heap: push")
    let input1 = 2
    print("  入力値: \(input1)")
    let output1 = minHeap.push(input1)
    print("  出力値: \(output1)")
    print("  現在のデータ: \(minHeap.get())")
    
    print("\nmin heap: push")
    let input2 = 15
    print("  入力値: \(input2)")
    let output2 = minHeap.push(input2)
    print("  出力値: \(output2)")
    print("  現在のデータ: \(minHeap.get())")
    
    print("\nmin heap: peek")
    let output3 = minHeap.peek()
    print("  出力値: \(String(describing: output3))")
    
    print("\nmin heap: pop")
    let output4 = minHeap.pop()
    print("  出力値: \(output4)")
    print("  現在のデータ: \(minHeap.get())")
    
    print("\nmin heap: pop")
    let output5 = minHeap.pop()
    print("  出力値: \(output5)")
    print("  現在のデータ: \(minHeap.get())")
    
    print("\nmin heap: get_index")
    let input3 = 3
    print("  入力値: \(input3)")
    let output6 = minHeap.getIndex(of: input3)
    print("  出力値: \(output6)")
    
    print("\nmin heap: get_index")
    let input4 = 100
    print("  入力値: \(input4)")
    let output7 = minHeap.getIndex(of: input4)
    print("  出力値: \(output7)")
    
    print("\nmin heap: is_empty")
    let output8 = minHeap.isEmpty()
    print("  出力値: \(output8)")
    
    print("\nmin heap: size")
    let output9 = minHeap.size()
    print("  出力値: \(output9)")
    
    print("\nmin heap: clear")
    let output10 = minHeap.clear()
    print("  出力値: \(output10)")
    print("  現在のデータ: \(minHeap.get())")
    
    print("\nmin heap: is_empty")
    let output11 = minHeap.isEmpty()
    print("  出力値: \(output11)")
    
    print("\nmax heap: new")
    let maxHeap = HeapData(isMinHeap: false)
    print("  現在のデータ: \(maxHeap.get())")
    
    print("\nmax heap: heapify")
    let input5 = [4, 10, 3, 5, 1]
    print("  入力値: \(input5)")
    let output12 = maxHeap.heapify(input5)
    print("  出力値: \(output12)")
    print("  現在のデータ: \(maxHeap.get())")
    
    print("\nmax heap: push")
    let input6 = 12
    print("  入力値: \(input6)")
    let output13 = maxHeap.push(input6)
    print("  出力値: \(output13)")
    print("  現在のデータ: \(maxHeap.get())")
    
    print("\nmax heap: push")
    let input7 = 0
    print("  入力値: \(input7)")
    let output14 = maxHeap.push(input7)
    print("  出力値: \(output14)")
    print("  現在のデータ: \(maxHeap.get())")
    
    print("\nmax heap: peek")
    let output15 = maxHeap.peek()
    print("  出力値: \(String(describing: output15))")
    
    print("\nmax heap: pop")
    let output16 = maxHeap.pop()
    print("  出力値: \(output16)")
    print("  現在のデータ: \(maxHeap.get())")
    
    print("\nmax heap: pop")
    let output17 = maxHeap.pop()
    print("  出力値: \(output17)")
    print("  現在のデータ: \(maxHeap.get())")
    
    print("\nmax heap: get_index")
    let input8 = 5
    print("  入力値: \(input8)")
    let output18 = maxHeap.getIndex(of: input8)
    print("  出力値: \(output18)")
    
    print("\nmax heap: get_index")
    let input9 = -10
    print("  入力値: \(input9)")
    let output19 = maxHeap.getIndex(of: input9)
    print("  出力値: \(output19)")
    
    print("\nmax heap: is_empty")
    let output20 = maxHeap.isEmpty()
    print("  出力値: \(output20)")
    
    print("\nmax heap: size")
    let output21 = maxHeap.size()
    print("  出力値: \(output21)")
    
    print("\nmax heap: clear")
    let output22 = maxHeap.clear()
    print("  出力値: \(output22)")
    print("  現在のデータ: \(maxHeap.get())")
    
    print("\nmax heap: is_empty")
    let output23 = maxHeap.isEmpty()
    print("  出力値: \(output23)")
    
    print("\nHeap TEST <----- end")
}

main()

File: data_structure/heap/src/HeapDemo.rs
// Rust
// データ構造: ヒープ (Heap)

use std::collections::BinaryHeap;
use std::cmp::Reverse;

pub struct HeapData<T: Ord> {
    data: BinaryHeap<T>,
    is_min_heap: bool,
}

impl<T: Ord + Clone> HeapData<T> {
    pub fn new(is_min_heap: bool) -> Self {
        HeapData {
            data: BinaryHeap::new(),
            is_min_heap,
        }
    }

    pub fn get(&self) -> Vec<T> {
        if self.is_min_heap {
            self.data.clone().into_sorted_vec().into_iter().rev().collect()
        } else {
            self.data.clone().into_sorted_vec()
        }
    }

    pub fn get_index(&self, item: &T) -> Option<usize> 
    where T: PartialEq {
        self.get().iter().position(|x| x == item)
    }

    pub fn get_value(&self, index: usize) -> Option<&T> {
        self.get().get(index)
    }

    pub fn heapify(&mut self, array: Vec<T>) -> bool {
        if self.is_min_heap {
            self.data = array.into_iter().map(Reverse).collect();
        } else {
            self.data = array.into_iter().collect();
        }
        true
    }

    pub fn push(&mut self, value: T) -> bool {
        if self.is_min_heap {
            self.data.push(Reverse(value));
        } else {
            self.data.push(value);
        }
        true
    }

    pub fn pop(&mut self) -> bool {
        if self.is_min_heap {
            self.data.pop();
        } else {
            self.data.pop();
        }
        true
    }

    pub fn peek(&self) -> Option<T> {
        if self.is_min_heap {
            self.data.peek().map(|x| x.0.clone())
        } else {
            self.data.peek().cloned()
        }
    }

    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    pub fn size(&self) -> usize {
        self.data.len()
    }

    pub fn clear(&mut self) -> bool {
        self.data.clear();
        true
    }
}

fn main() {
    println!("Heap TEST -----> start");

    println!("\nmin heap: new");
    let mut min_heap = HeapData::<i32>::new(true);
    println!("  現在のデータ: {:?}", min_heap.get());

    println!("\nmin heap: heapify");
    let input = vec![4, 10, 3, 5, 1];
    println!("  入力値: {:?}", input);
    let output = min_heap.heapify(input);
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", min_heap.get());

    println!("\nmin heap: push");
    let input = 2;
    println!("  入力値: {}", input);
    let output = min_heap.push(input);
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", min_heap.get());

    println!("\nmin heap: push");
    let input = 15;
    println!("  入力値: {}", input);
    let output = min_heap.push(input);
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", min_heap.get());

    println!("\nmin heap: peek");
    let output = min_heap.peek();
    println!("  出力値: {:?}", output);

    println!("\nmin heap: pop");
    let output = min_heap.pop();
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", min_heap.get());

    println!("\nmin heap: pop");
    let output = min_heap.pop();
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", min_heap.get());

    println!("\nmin heap: get_index");
    let input = 3;
    println!("  入力値: {}", input);
    let output = min_heap.get_index(&input);
    println!("  出力値: {:?}", output);

    println!("\nmin heap: get_index");
    let input = 100;
    println!("  入力値: {}", input);
    let output = min_heap.get_index(&input);
    println!("  出力値: {:?}", output);

    println!("\nmin heap: is_empty");
    let output = min_heap.is_empty();
    println!("  出力値: {}", output);

    println!("\nmin heap: size");
    let output = min_heap.size();
    println!("  出力値: {}", output);

    println!("\nmin heap: clear");
    let output = min_heap.clear();
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", min_heap.get());

    println!("\nmin heap: is_empty");
    let output = min_heap.is_empty();
    println!("  出力値: {}", output);

    println!("\nmax heap: new");
    let mut max_heap = HeapData::<i32>::new(false);
    println!("  現在のデータ: {:?}", max_heap.get());

    println!("\nmax heap: heapify");
    let input = vec![4, 10, 3, 5, 1];
    println!("  入力値: {:?}", input);
    let output = max_heap.heapify(input);
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", max_heap.get());

    println!("\nmax heap: push");
    let input = 12;
    println!("  入力値: {}", input);
    let output = max_heap.push(input);
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", max_heap.get());

    println!("\nmax heap: push");
    let input = 0;
    println!("  入力値: {}", input);
    let output = max_heap.push(input);
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", max_heap.get());

    println!("\nmax heap: peek");
    let output = max_heap.peek();
    println!("  出力値: {:?}", output);

    println!("\nmax heap: pop");
    let output = max_heap.pop();
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", max_heap.get());

    println!("\nmax heap: pop");
    let output = max_heap.pop();
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", max_heap.get());

    println!("\nmax heap: get_index");
    let input = 5;
    println!("  入力値: {}", input);
    let output = max_heap.get_index(&input);
    println!("  出力値: {:?}", output);

    println!("\nmax heap: get_index");
    let input = -10;
    println!("  入力値: {}", input);
    let output = max_heap.get_index(&input);
    println!("  出力値: {:?}", output);

    println!("\nmax heap: is_empty");
    let output = max_heap.is_empty();
    println!("  出力値: {}", output);

    println!("\nmax heap: size");
    let output = max_heap.size();
    println!("  出力値: {}", output);

    println!("\nmax heap: clear");
    let output = max_heap.clear();
    println!("  出力値: {:?}", output);
    println!("  現在のデータ: {:?}", max_heap.get());

    println!("\nmax heap: is_empty");
    let output = max_heap.is_empty();
    println!("  出力値: {}", output);

    println!("\nHeap TEST <----- end");
}
